vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,322 @@
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBlockSize => write!(f, "invalid block size count"),
+            InvalidEntryCount => write!(f, "invalid metadata entry count"),
+            InvalidLogicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidMetadataItem => write!(f, "invalid metadata ID"),
+            InvalidMetadataLength => write!(f, "invalid metadata length"),
+            InvalidMetadataSign => write!(f, "metadata sign doesn't match"),
+            InvalidPhysicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidValue => write!(f, "invalid value"),
+            MissingMetadata => write!(f, "not all required metadata found"),
+            ReadMetadata(e) => write!(f, "failed to read metadata headers: {}", e),
+            UnsupportedFlag => write!(f, "this implementation support this metadata flag"),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct DiskSpec {
+    pub disk_id: u128,
+    pub image_size: u64,
+    pub block_size: u32,
+    pub has_parent: bool,
+    pub sectors_per_block: u32,
+    pub virtual_disk_size: u64,
+    pub logical_sector_size: u32,
+    pub physical_sector_size: u32,
+    pub chunk_ratio: u64,
+    pub total_sectors: u64,
+}
+
+struct MetadataTableHeader {
+    _signature: u64,
+    _reserved: u16,
+    entry_count: u16,
+    _reserved2: [u8; 20],
+}
+
+impl MetadataTableHeader {
+    pub fn new(buffer: &[u8]) -> Result<MetadataTableHeader> {
+        let signature = LittleEndian::read_u64(&buffer[0..8]);
+        if signature != METADATA_SIGN {
+            return Err(Error::InvalidMetadataSign);
+        }
+
+        let _reserved = LittleEndian::read_u16(&buffer[8..10]);
+        let entry_count = LittleEndian::read_u16(&buffer[10..12]);
+        if entry_count > METADATA_MAX_ENTRIES {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        Ok(MetadataTableHeader {
+            _signature: signature,
+            _reserved,
+            entry_count,
+            _reserved2: buffer[12..32]
+                .try_into()
+                .expect("slice with incorrect length"),
+        })
+    }
+}
+
+pub struct MetadataTableEntry {
+    item_id: u128,
+    offset: u32,
+    length: u32,
+    flag_bits: u32,
+    _reserved: u32,
+}
+
+impl MetadataTableEntry {
+    /// Parse one metadata entry from the buffer
+    fn new(buffer: &[u8]) -> Result<MetadataTableEntry> {
+        let metadata_table_entry = MetadataTableEntry {
+            item_id: LittleEndian::read_u128(&buffer[0..16]),
+            offset: LittleEndian::read_u32(&buffer[16..20]),
+            length: LittleEndian::read_u32(&buffer[20..24]),
+            flag_bits: LittleEndian::read_u32(&buffer[24..28]),
+            _reserved: LittleEndian::read_u32(&buffer[28..32]),
```

Comment:
Don't initialize `_reserved`, just use the default implementation of `MetadataTableEntry`.